lms/services/course.py

```diff
@@ -51,61 +57,43 @@ def upsert(self, context_id, name, extra, settings=None) -> Course:
         :param extra: Additional LMS specific values
         :param settings: A dict of settings for the course
         """
-        authority_provided_id = self._get_authority_provided_id(context_id)
-
-        course = self._get_by_authority_provided_id(authority_provided_id)
-        if not course:
-            course = Course(
-                application_instance_id=self._application_instance.id,
-                authority_provided_id=authority_provided_id,
-                lms_id=context_id,
-                lms_name=name,
-                settings=settings or self._new_course_settings(authority_provided_id),
-                extra=extra,
-            )
-            self._db.add(course)
-
-        # Update any values that might have changed
-        course.lms_name = name
-        course.extra = extra
 
-        return course
-
-    def _get_authority_provided_id(self, context_id):
-        return GroupingService.generate_authority_provided_id(
-            tool_consumer_instance_guid=self._application_instance.tool_consumer_instance_guid,
-            lms_id=context_id,
-            parent=None,
+        return self._grouping_service.upsert_groupings(
+            [
+                {
+                    "lms_id": context_id,
+                    "lms_name": name,
+                    "extra": extra,
+                    "settings": settings or self._new_course_settings(context_id),
+                }
+            ],
             type_=Grouping.Type.COURSE,
-        )
+        )[0]
 
-    def _get_by_authority_provided_id(self, authority_provided_id):
-        return (
-            self._db.query(Course)
-            .filter_by(
-                application_instance=self._application_instance,
-                authority_provided_id=authority_provided_id,
-            )
-            .one_or_none()
+    def _get_authority_provided_id(self, context_id):
+        return self._grouping_service.get_authority_provided_id(
+            lms_id=context_id, type_=Grouping.Type.COURSE
         )
 
-    def _new_course_settings(self, authority_provided_id):
+    def _new_course_settings(self, context_id):
         # By default we'll make our course setting have the same settings
         # as the application instance
         course_settings = deepcopy(self._application_instance.settings)
 
         # Unless! The group was pre-sections, and we've just seen it for the
         # first time in which case turn sections off
         if course_settings.get("canvas", "sections_enabled") and self._is_pre_sections(
-            authority_provided_id
+            context_id
         ):
             course_settings.set("canvas", "sections_enabled", False)
 
         return course_settings
 
-    def _is_pre_sections(self, authority_provided_id):
+    def _is_pre_sections(self, context_id):
         return bool(
```

Comment:
I wonder how often this is true. Probably something to look at and probably remove in another pass.

(It's not a lot of code though, doesn't get in the way).